crates/farming/src/lib.rs

```diff
@@ -0,0 +1,352 @@
+
+    pub(crate) type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+
+    pub(crate) type CurrencyIdOf<T> =
+        <<T as Config>::MultiCurrency as MultiCurrency<<T as frame_system::Config>::AccountId>>::CurrencyId;
+
+    pub(crate) type BalanceOf<T> = <<T as Config>::MultiCurrency as MultiCurrency<AccountIdOf<T>>>::Balance;
+
+    pub(crate) type RewardScheduleOf<T> = RewardSchedule<BalanceOf<T>>;
+
+    /// ## Configuration
+    /// The pallet's configuration trait.
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// The overarching event type.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        /// The farming pallet id, used for deriving pool accounts.
+        #[pallet::constant]
+        type FarmingPalletId: Get<PalletId>;
+
+        /// The treasury pallet id, used for deriving its sovereign account ID.
+        #[pallet::constant]
+        type TreasuryPalletId: Get<PalletId>;
+
+        /// The period to accrue rewards.
+        #[pallet::constant]
+        type RewardPeriod: Get<Self::BlockNumber>;
+
+        /// Reward pools to track stake.
+        type RewardPools: RewardsApi<
+            CurrencyIdOf<Self>, // pool id is the lp token
+            AccountIdOf<Self>,
+            BalanceOf<Self>,
+            CurrencyId = CurrencyIdOf<Self>,
+        >;
+
+        /// Currency handler to transfer tokens.
+        type MultiCurrency: MultiReservableCurrency<AccountIdOf<Self>, CurrencyId = CurrencyId>;
+
+        /// Weight information for the extrinsics.
+        type WeightInfo: WeightInfo;
+    }
+
+    // The pallet's events
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(crate) fn deposit_event)]
+    pub enum Event<T: Config> {
+        RewardScheduleUpdated {
+            pool_currency_id: CurrencyIdOf<T>,
+            reward_currency_id: CurrencyIdOf<T>,
+            period_count: u32,
+            per_period: BalanceOf<T>,
+        },
+        RewardDistributed {
+            pool_currency_id: CurrencyIdOf<T>,
+            reward_currency_id: CurrencyIdOf<T>,
+            amount: BalanceOf<T>,
+        },
+        RewardClaimed {
+            account_id: AccountIdOf<T>,
+            pool_currency_id: CurrencyIdOf<T>,
+            reward_currency_id: CurrencyIdOf<T>,
+            amount: BalanceOf<T>,
+        },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        ScheduleNotFound,
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+        fn on_initialize(now: T::BlockNumber) -> Weight {
+            if now % T::RewardPeriod::get() == Zero::zero() {
+                let mut count: u32 = 0;
+                // collect first to avoid modifying in-place
+                let schedules = RewardSchedules::<T>::iter().collect::<Vec<_>>();
+                for (pool_currency_id, reward_currency_id, mut reward_schedule) in schedules.into_iter() {
```

Comment:
For backlog, we want to check this complexity/impact on our block times if we have many pools (say 50 or 100 pools).

Reply:
@gregdhill did you add an issue for that?

Reply:
https://github.com/interlay/interbtc/issues/857